experimental_platforms mac os to linux does not work

Cross compiling with release 0.7.0 --experimental_platforms does not work on mac to produce linux binaries.

I have create a sample repo that fails with the command `bazel build --experimental_platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 //:logrusbazel`

https://github.com/margic/logrusbazel

Command output 
```
 bazel build --experimental_platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 //:logrusbazel
INFO: Found 1 target...
ERROR: /Users/paul.crofts/Documents/go/src/github.com/margic/logrusbazel/vendor/golang.org/x/sys/unix/BUILD.bazel:3:1: GoCompile vendor/golang.org
/x/sys/unix/~pure_stripped~go_default_library~/~partial.a failed (Exit 1).
bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_flock.cgo1.go:16:42: undefined: Flock_t
bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_sockcmsg_unix.cgo1.go:37:18: undefined: Cmsghdr
bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_sockcmsg_unix.cgo1.go:43:9: undefined: Cmsghdr
2017/11/09 13:36:46 error running compiler: exit status 1
Target //:logrusbazel failed to build
Use --verbose_failures to see the command lines o
```

Full command output with errors
```
bazel build --verbose_failures --experimental_platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 //:logrusbazel
INFO: Found 1 target...
ERROR: /Users/paul.crofts/Documents/go/src/github.com/margic/logrusbazel/vendor/golang.org/x/sys/unix/BUILD.bazel:3:1: GoCompile vendor/golang.org
/x/sys/unix/~pure_stripped~go_default_library~/~partial.a failed (Exit 1): compile failed: error executing command
  (cd /private/var/tmp/_bazel_paul.crofts/fc93648e8fa8d6c7ba0f8e338fa9b101/execroot/__main__ && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/compile -go bazel-out/host/bin/external/go_stdlib_linux_amd64_pure/bin/go -root_file bazel-out/host/bin/external/go_stdlib_linux_amd64_pure/ROOT -goos linux -goarch amd64 '-cgo=0' -package_list external/go_sdk/packages.txt -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_env_unset.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_constants.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_dev_darwin.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_dirent.cgo1.go -srcbazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_endian_little.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_env_unix.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_flock.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_pagesize_unix.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_race0.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_sockcmsg_unix.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_str.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_syscall.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_syscall_bsd.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_syscall_darwin.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_syscall_darwin_amd64.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_syscall_unix.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_syscall_unix_gc.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_timestruct.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_zerrors_darwin_amd64.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_zsyscall_darwin_amd64.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_zsysnum_darwin_amd64.cgo1.go -src bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_ztypes_darwin_amd64.cgo1.go -o 'bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/~pure_stripped~go_default_library~/~partial.a' -trimpath . -I . -- -p golang.org/x/sys/unix bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/_cgo_gotypes.go bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/_cgo_import.go)

Use --sandbox_debug to see verbose messages from the sandbox.
bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_flock.cgo1.go:16:42: undefined: Flock_t
bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_sockcmsg_unix.cgo1.go:37:18: undefined: Cmsghdr
bazel-out/darwin_x86_64-fastbuild/bin/vendor/golang.org/x/sys/unix/go_default_library.cgo.dir/vendor_golang.org_x_sys_unix_sockcmsg_unix.cgo1.go:43:9: undefined: Cmsghdr
2017/11/09 13:39:09 error running compiler: exit status 1
Target //:logrusbazel failed to build
INFO: Elapsed time: 5.296s, Critical Path: 5.02s
```

--- a/go/private/actions/archive.bzl
+++ b/go/private/actions/archive.bzl
@@ -35,6 +35,7 @@ def emit_archive(ctx, go_toolchain, mode=None, importpath=None, source=None, imp
   if source == None: fail("source is a required parameter")
   if mode == None: fail("mode is a required parameter")
 
+  source = sources.filter(ctx, source, mode)
 
   cover_vars = []
   if ctx.configuration.coverage_enabled:

--- a/go/private/providers.bzl
+++ b/go/private/providers.bzl
@@ -39,7 +39,7 @@ def _merge_runfiles(a, b):
   if not b: return a
   return a.merge(b)
 
-def _source_build_entry(srcs = [], deps = [], gc_goopts=[], runfiles=None, cgo_deps=[], cgo_exports=[], cgo_archive=None, want_coverage = False, source = None):
+def _source_build_entry(srcs = [], deps = [], gc_goopts=[], runfiles=None, cgo_deps=[], cgo_exports=[], cgo_archive=None, want_coverage = False, source = None, exclude = None):
   """Creates a new GoSource from a collection of values and an optional GoSourceList to merge in."""
   for e in (source.entries if source else []):
     srcs = srcs + e.srcs
@@ -62,6 +62,7 @@ def _source_build_entry(srcs = [], deps = [], gc_goopts=[], runfiles=None, cgo_d
       cgo_exports = cgo_exports,
       cgo_archive = cgo_archive,
       want_coverage = want_coverage,
+      exclude = exclude,
   )
 
 def _source_new(**kwargs):
@@ -79,9 +80,13 @@ def _source_flatten(ctx, source):
   """Flattens a GoSourceList to a single GoSource ready for use."""
   return _source_build_entry(source = source)
 
+def _source_filter(ctx, source, mode):
+  return GoSourceList(entries = [s for s in source.entries if not (s.exclude and s.exclude(ctx, mode))])
+
 sources = struct(
   new = _source_new,
   merge = _source_merge,
   flatten = _source_flatten,
+  filter = _source_filter,
 )
 """sources holds the functions for manipulating GoSourceList providers."""

--- a/go/private/rules/aspect.bzl
+++ b/go/private/rules/aspect.bzl
@@ -79,10 +79,10 @@ go_archive_aspect = aspect(
     _go_archive_aspect_impl,
     attr_aspects = ["deps", "embed"],
     attrs = {
-        "pure": attr.string(values=["on", "off", "auto"], default="auto"),
-        "static": attr.string(values=["on", "off", "auto"], default="auto"),
-        "msan": attr.string(values=["on", "off", "auto"], default="auto"),
-        "race": attr.string(values=["on", "off", "auto"], default="auto"),
+        "pure": attr.string(values=["on", "off", "auto"]),
+        "static": attr.string(values=["on", "off", "auto"]),
+        "msan": attr.string(values=["on", "off", "auto"]),
+        "race": attr.string(values=["on", "off", "auto"]),
     },
     toolchains = ["@io_bazel_rules_go//go:toolchain"],
 )

--- a/go/private/rules/cgo.bzl
+++ b/go/private/rules/cgo.bzl
@@ -149,6 +149,7 @@ def _cgo_codegen_impl(ctx):
       ),
       OutputGroupInfo(
           go_files = to_set(go_outs),
+          input_go_files = to_set(source.go),
           c_files = sets.union(c_outs, source.headers),
           main_c = to_set([cgo_main]),
       ),
@@ -220,26 +221,42 @@ Args:
   out: Destination of the generated codes.
 """
 
+def _pure(ctx, mode):
+    return mode.pure
+
+def _not_pure(ctx, mode):
+    return not mode.pure
+
 def _cgo_collect_info_impl(ctx):
   codegen = ctx.attr.codegen[_CgoCodegen]
   runfiles = ctx.runfiles(collect_data = True)
   runfiles = runfiles.merge(ctx.attr.codegen.data_runfiles)
   return [
       DefaultInfo(files = depset(), runfiles = runfiles),
-      sources.new(
-          srcs = ctx.files.gen_go_srcs,
-          runfiles = runfiles,
-          cgo_deps = ctx.attr.codegen[_CgoCodegen].deps,
-          cgo_exports = ctx.attr.codegen[_CgoCodegen].exports,
-          cgo_archive = _select_archive(ctx.files.lib),
-          want_coverage = ctx.coverage_instrumented(), #TODO: not all sources?
-      ),
+      sources.merge([
+          sources.new(
+              srcs = ctx.files.gen_go_srcs,
+              runfiles = runfiles,
+              cgo_deps = ctx.attr.codegen[_CgoCodegen].deps,
+              cgo_exports = ctx.attr.codegen[_CgoCodegen].exports,
+              cgo_archive = _select_archive(ctx.files.lib),
+              want_coverage = ctx.coverage_instrumented(), #TODO: not all sources?
+              exclude = _pure,
+          ),
+          sources.new(
+              srcs = ctx.files.input_go_srcs,
+              runfiles = runfiles,
+              want_coverage = ctx.coverage_instrumented(), #TODO: not all sources?
+              exclude = _not_pure,
+          ),
+      ]),
   ]
 
 _cgo_collect_info = rule(
     _cgo_collect_info_impl,
     attrs = {
         "codegen": attr.label(mandatory = True, providers = [_CgoCodegen]),
+        "input_go_srcs": attr.label_list(mandatory = True, allow_files = [".go"]),
         "gen_go_srcs": attr.label_list(mandatory = True, allow_files = [".go"]),
         "lib": attr.label(mandatory = True, providers = ["cc"]),
         "_go_toolchain_flags": attr.label(default=Label("@io_bazel_rules_go//go/private:go_toolchain_flags")),
@@ -277,6 +294,14 @@ def setup_cgo_library(name, srcs, cdeps, copts, clinkopts):
       name = select_go_files,
       srcs = [cgo_codegen_name],
       output_group = "go_files",
+      visibility = ["//visibility:private"],
+  )
+
+  select_input_go_files = name + ".select_input_go_files"
+  native.filegroup(
+      name = select_input_go_files,
+      srcs = [cgo_codegen_name],
+      output_group = "input_go_files",
       visibility = ["//visibility:private"],
   )
 
@@ -349,6 +374,9 @@ def setup_cgo_library(name, srcs, cdeps, copts, clinkopts):
   _cgo_collect_info(
       name = cgo_embed_name,
       codegen = cgo_codegen_name,
+      input_go_srcs = [
+          select_input_go_files,
+      ],
       gen_go_srcs = [
           select_go_files,
           cgo_import_name,

--- a/tests/cgo_pure/BUILD.bazel
+++ b/tests/cgo_pure/BUILD.bazel
@@ -30,5 +30,4 @@ go_test(
     x_defs = {
         "github.com/bazelbuild/rules_go/tests/cgo_pure.Expect": "1",
     },
-    tags = ["manual"], #TODO: remove this once it passes, see #1009
 )

